tensorwaves/physics/helicityformalism/amplitude.py

```diff
@@ -1,246 +1,402 @@
+        return IntensityTF(
+            self._create_intensity(recipe["Intensity"]), self._parameters
         )
 
+    def _create_intensity(self, recipe: dict) -> Callable:
+        intensity_class = recipe["Class"]
+        logging.debug("creating %s", intensity_class)
+
+        # this dict could be used for builder injections later on
+        basic_intensity_builders: Dict[str, Callable] = {
+            "IncoherentIntensity": _IncoherentIntensity,
+            "CoherentIntensity": _CoherentIntensity,
+            "StrengthIntensity": _StrengthIntensity,
+        }
+        intensity_kinematics_builders = {
+            "NormalizedIntensity": _NormalizedIntensity,
+        }
+
+        if intensity_class in basic_intensity_builders:
+            return basic_intensity_builders[intensity_class](recipe, self)
+        if intensity_class in intensity_kinematics_builders:
+            return intensity_kinematics_builders[intensity_class](
+                recipe, self, self._kinematics
+            )
 
-def _create_amplitude(recipe: dict):
-    amplitude_class = recipe["Class"]
-    logging.info("creating", amplitude_class)
-
-    known_amplitude_builders = {
-        "CoefficientAmplitude": _CoefficientAmplitude,
-        "SequentialAmplitude": _SequentialAmplitude,
-        "HelicityDecay": _HelicityDecay,
-    }
-    if amplitude_class in known_amplitude_builders:
-        return known_amplitude_builders[amplitude_class](recipe)
-    else:
-        logging.error(
-            "unknown intensity " + str(amplitude_class) + "! Skipping"
+        raise Exception("Unknown intensity {}!".format(intensity_class))
+
+    def _create_amplitude(self, recipe: dict) -> Callable:
+        amplitude_class = recipe["Class"]
+        logging.debug("creating %s", amplitude_class)
+
+        basic_amplitude_builders = {
+            "CoefficientAmplitude": _CoefficientAmplitude,
+            "SequentialAmplitude": _SequentialAmplitude,
+        }
+        general_amplitude_builders = {
+            "HelicityDecay": _HelicityDecay,
+        }
+
+        if amplitude_class in basic_amplitude_builders:
+            return basic_amplitude_builders[amplitude_class](recipe, self)
+        if amplitude_class in general_amplitude_builders:
+            return general_amplitude_builders[amplitude_class](
+                recipe, self, self._kinematics, self._particles, self._dynamics
+            )
+
+        raise Exception("Unknown amplitude {}!".format(amplitude_class))
+
+    def _register_parameter(self, name: str, value: float) -> tf.Variable:
+        if name not in self._parameters:
+            self._parameters[name] = tf.Variable(
+                value, name=name, dtype=tf.float64
+            )
+        return self._parameters[name]
+
+    def _get_parameter(self, name: str) -> tf.Variable:
+        if name not in self._parameters:
+            raise Exception(
+                "Parameter {} not registered! Your recipe file is"
```

Comment:
Since we have given up support, we can make use of [literal string interpolation](https://www.python.org/dev/peps/pep-0498/), which is nicer to read.

Reply:
https://github.com/ComPWA/tensorwaves/pull/56/commits/5f7937d96cda5d51460ec1900d0d84e13d2ddf24